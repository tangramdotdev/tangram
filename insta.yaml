test:
  auto_review: true
  runner: nextest
