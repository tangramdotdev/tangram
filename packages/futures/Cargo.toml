[package]
name = "tangram_futures"

description = { workspace = true }
edition = { workspace = true }
homepage = { workspace = true }
license = { workspace = true }
publish = false
repository = { workspace = true }
rust-version = { workspace = true }
version = { workspace = true }

[lints]
workspace = true

[dependencies]
dashmap = { workspace = true }
futures = { workspace = true }
tokio = { workspace = true }
tokio-util = { workspace = true }
