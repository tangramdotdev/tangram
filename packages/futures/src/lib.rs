pub mod future;
pub mod read;
pub mod stream;
pub mod task;
pub mod write;
