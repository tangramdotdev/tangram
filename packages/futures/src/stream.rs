use self::attach::Attach;
use self::take_while_inclusive::TakeWhileInclusive;
use futures::{Future, Stream, StreamExt as _, TryStream, TryStreamExt as _};

pub mod attach;
pub mod take_while_inclusive;

pub trait Ext: Stream {
	fn attach<T>(self, value: T) -> Attach<Self, T>
	where
		Self: Sized,
	{
		Attach::new(self, value)
	}

	fn last(mut self) -> impl Future<Output = Option<Self::Item>>
	where
		Self: Sized + Unpin,
	{
		async move {
			let mut last = None;
			while let Some(item) = self.next().await {
				last = Some(item);
			}
			last
		}
	}

	fn take_while_inclusive<F, Fut>(self, predicate: F) -> TakeWhileInclusive<Self, F, Fut>
	where
		Self: Sized,
		F: FnMut(&Self::Item) -> Fut,
		Fut: Future<Output = bool>,
	{
		TakeWhileInclusive::new(self, predicate)
	}
}

impl<S> Ext for S where S: Stream {}

pub trait TryExt: TryStream {
	fn try_last(mut self) -> impl Future<Output = Result<Option<Self::Ok>, Self::Error>>
	where
		Self: Sized + Unpin,
	{
		async move {
			let mut last = None;
			while let Some(item) = self.try_next().await? {
				last = Some(item);
			}
			Ok(last)
		}
	}
}

impl<S> TryExt for S where S: TryStream {}
