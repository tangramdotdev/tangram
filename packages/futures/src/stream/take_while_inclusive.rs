use futures::{Future, Stream};
use pin_project::pin_project;
use std::{
	pin::Pin,
	task::{Context, Poll},
};

#[pin_project]
pub struct TakeWhileInclusive<S, F, Fut>
where
	S: Stream,
{
	#[pin]
	stream: S,
	predicate: F,
	#[pin]
	pending_predicate: Option<Fut>,
	pending_item: Option<S::Item>,
	done: bool,
}

impl<S, F, Fut> TakeWhileInclusive<S, F, Fut>
where
	S: Stream,
	F: FnMut(&S::Item) -> Fut,
	Fut: Future<Output = bool>,
{
	pub fn new(stream: S, predicate: F) -> Self {
		Self {
			stream,
			predicate,
			pending_predicate: None,
			pending_item: None,
			done: false,
		}
	}
}

impl<S, F, Fut> Stream for TakeWhileInclusive<S, F, Fut>
where
	S: Stream,
	F: FnMut(&S::Item) -> Fut,
	Fut: Future<Output = bool>,
{
	type Item = S::Item;

	fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
		let mut this = self.project();

		if *this.done {
			return Poll::Ready(None);
		}

		loop {
			if let Some(pending) = this.pending_predicate.as_mut().as_pin_mut() {
				match pending.poll(cx) {
					Poll::Ready(true) => {
						let item = this.pending_item.take().unwrap();
						this.pending_predicate.set(None);
						return Poll::Ready(Some(item));
					},
					Poll::Ready(false) => {
						let item = this.pending_item.take().unwrap();
						*this.done = true;
						return Poll::Ready(Some(item));
					},
					Poll::Pending => return Poll::Pending,
				}
			}

			match this.stream.as_mut().poll_next(cx) {
				Poll::Ready(Some(item)) => {
					let future = (this.predicate)(&item);
					*this.pending_item = Some(item);
					this.pending_predicate.set(Some(future));
				},
				Poll::Ready(None) => {
					*this.done = true;
					return Poll::Ready(None);
				},
				Poll::Pending => return Poll::Pending,
			}
		}
	}
}

#[cfg(test)]
mod tests {
	use crate::stream::Ext as _;
	use futures::{future, stream, StreamExt as _};

	#[tokio::test]
	async fn take_while_inclusive() {
		let values = stream::iter([1, 2, 3, 4, 5])
			.take_while_inclusive(|value| future::ready(*value != 4))
			.collect::<Vec<_>>()
			.await;
		assert_eq!(values, [1, 2, 3, 4]);
	}
}
