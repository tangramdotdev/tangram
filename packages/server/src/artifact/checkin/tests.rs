use crate::{Config, Server};
use futures::FutureExt as _;
use insta::assert_snapshot;
use std::{future::Future, panic::AssertUnwindSafe, path::Path, pin::pin};
use tangram_client as tg;
use tangram_futures::stream::TryStreamExt as _;
use tangram_temp::{self as temp, artifact, Temp};

type Result = std::result::Result<(), Box<dyn std::error::Error>>;

#[tokio::test]
async fn basic() -> Result {
	let artifact = artifact!({
		"directory": {
			"README.md": "Hello, World!",
		}
	});
	let path = "directory";
	test_checkin(artifact, path, |server, artifact| async { Ok(()) }).await
}

async fn test_checkin<F, Fut>(artifact: temp::Artifact, path: &str, assertions: F) -> Result
where
	F: FnOnce(&Server, &tg::Artifact) -> Fut,
	Fut: Future<Output = Result>,
{
	test_server(|server| async move {
		let directory = Temp::new();
		artifact.to_path(directory.as_ref()).await?;
		let arg = tg::artifact::checkin::Arg {
			destructive: false,
			deterministic: false,
			ignore: true,
			locked: false,
			path: directory.as_ref().join(path),
		};
		let stream = server.check_in_artifact(arg).await?;
		let output = pin!(stream)
			.try_last()
			.await?
			.and_then(|event| event.try_unwrap_output().ok())
			.ok_or_else(|| tg::error!("stream ended without output"))?;
		let artifact = tg::Artifact::with_id(output.artifact);
		tg::Object::from(artifact.clone())
			.load_recursive(&server)
			.await?;
		let output = tg::Value::from(artifact.clone()).print(tg::value::print::Options {
			recursive: true,
			style: tg::value::print::Style::Pretty { indentation: "\t" },
		});
		assert_snapshot!(output);
		Ok(())
	})
	.await
}

async fn test_server<F, Fut>(f: F) -> Result
where
	F: FnOnce(Server) -> Fut,
	Fut: Future<Output = Result>,
{
	let temp = Temp::new();
	let options = Config::with_path(temp.path().to_owned());
	let server = Server::start(options).await?;
	let result = AssertUnwindSafe((f)(server.clone())).catch_unwind().await;
	server.stop();
	server.wait().await;
	result.unwrap()
}
