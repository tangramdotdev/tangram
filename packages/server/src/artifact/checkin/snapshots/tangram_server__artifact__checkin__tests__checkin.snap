---
source: packages/server/src/artifact/checkin/tests.rs
expression: output
---
tg.directory({
	"README.md": tg.file({
		"contents": tg.leaf("Hello, World!"),
	}),
})
